File: app/backend/routers/admin_mfa.py
"""
/api/admin/mfa — 管理者 TOTP MFA エンドポイント

- GET  /status         MFA 設定状態を返す
- POST /setup          TOTP シークレット生成 + QR コード (base64)
- POST /setup/verify   セットアップ確認 → セッショントークン発行
- POST /verify         ログイン時 TOTP 検証 → セッショントークン発行
- GET  /session        セッショントークン検証
"""

import hashlib
import io
import base64
import secrets
import logging
from datetime import datetime, timezone, timedelta

import pyotp
import qrcode
from fastapi import APIRouter, Depends, HTTPException, Header, Request

import main
from routers.admin import require_admin

logger = logging.getLogger(__name__)

router = APIRouter()

_SESSION_DURATION_HOURS = 24
_ISSUER_NAME = "OpenRegimeAdmin"


def _hash_token(token: str) -> str:
    """トークンを SHA-256 ハッシュ化"""
    return hashlib.sha256(token.encode()).hexdigest()


def _get_user_email(supabase, user_id: str) -> str:
    """user_id からメールアドレスを取得"""
    result = (
        supabase.table("users")
        .select("email")
        .eq("id", user_id)
        .limit(1)
        .execute()
    )
    return result.data[0]["email"] if result.data else "admin"


def _create_session_token(supabase, user_id: str) -> dict:
    """MFA セッショントークンを生成・DB 保存"""
    token = secrets.token_hex(32)
    token_hash = _hash_token(token)
    expires_at = datetime.now(timezone.utc) + timedelta(hours=_SESSION_DURATION_HOURS)

    supabase.table("admin_mfa_sessions").insert({
        "user_id": user_id,
        "token_hash": token_hash,
        "expires_at": expires_at.isoformat(),
    }).execute()

    return {
        "token": token,
        "expires_at": expires_at.isoformat(),
    }


# ============================================================
# MFA Status
# ============================================================

@router.get("/status")
async def mfa_status(admin_id: str = Depends(require_admin)):
    """MFA の設定状態を返す"""
    supabase = main.get_supabase()
    result = (
        supabase.table("admin_mfa")
        .select("enabled")
        .eq("user_id", admin_id)
        .limit(1)
        .execute()
    )

    if not result.data:
        return {"mfa_enabled": False, "mfa_setup": False}

    return {
        "mfa_enabled": result.data[0]["enabled"],
        "mfa_setup": True,
    }


# ============================================================
# MFA Setup
# ============================================================

@router.post("/setup")
async def mfa_setup(admin_id: str = Depends(require_admin)):
    """TOTP シークレット生成 + QR コード返却"""
    supabase = main.get_supabase()

    # 既に有効な MFA がある場合は拒否
    existing = (
        supabase.table("admin_mfa")
        .select("enabled")
        .eq("user_id", admin_id)
        .limit(1)
        .execute()
    )
    if existing.data and existing.data[0]["enabled"]:
        raise HTTPException(status_code=409, detail="MFA already enabled")

    # TOTP シークレット生成
    secret = pyotp.random_base32()
    email = _get_user_email(supabase, admin_id)
    totp = pyotp.TOTP(secret)
    provisioning_uri = totp.provisioning_uri(name=email, issuer_name=_ISSUER_NAME)

    # QR コード生成 (base64 PNG)
    img = qrcode.make(provisioning_uri)
    buf = io.BytesIO()
    img.save(buf, format="PNG")
    qr_base64 = base64.b64encode(buf.getvalue()).decode()

    # DB に保存（未確認状態: enabled=False）
    if existing.data:
        supabase.table("admin_mfa").update({
            "secret_enc": secret,
            "enabled": False,
            "updated_at": datetime.now(timezone.utc).isoformat(),
        }).eq("user_id", admin_id).execute()
    else:
        supabase.table("admin_mfa").insert({
            "user_id": admin_id,
            "secret_enc": secret,
            "enabled": False,
        }).execute()

    return {
        "secret": secret,
        "qr_code": f"data:image/png;base64,{qr_base64}",
        "provisioning_uri": provisioning_uri,
    }


# ============================================================
# MFA Setup Verify
# ============================================================

@router.post("/setup/verify")
async def mfa_setup_verify(
    request: Request,
    admin_id: str = Depends(require_admin),
):
    """セットアップ確認: 6桁コードで検証 → MFA 有効化 + トークン発行"""
    supabase = main.get_supabase()
    body = await request.json()
    code = str(body.get("code", "")).strip()

    if not code or len(code) != 6:
        raise HTTPException(status_code=400, detail="6-digit code required")

    # シークレット取得
    result = (
        supabase.table("admin_mfa")
        .select("secret_enc, enabled")
        .eq("user_id", admin_id)
        .limit(1)
        .execute()
    )
    if not result.data:
        raise HTTPException(status_code=404, detail="MFA setup not found. Call /setup first.")

    if result.data[0]["enabled"]:
        raise HTTPException(status_code=409, detail="MFA already enabled")

    secret = result.data[0]["secret_enc"]
    totp = pyotp.TOTP(secret)

    if not totp.verify(code, valid_window=1):
        raise HTTPException(status_code=401, detail="Invalid code")

    # MFA を有効化
    supabase.table("admin_mfa").update({
        "enabled": True,
        "updated_at": datetime.now(timezone.utc).isoformat(),
    }).eq("user_id", admin_id).execute()

    # セッショントークン発行
    session = _create_session_token(supabase, admin_id)

    logger.info(f"MFA enabled for user {admin_id}")

    return {
        "status": "mfa_enabled",
        **session,
    }


# ============================================================
# MFA Verify (login)
# ============================================================

@router.post("/verify")
async def mfa_verify(
    request: Request,
    admin_id: str = Depends(require_admin),
):
    """ログイン時の TOTP 検証 → セッショントークン発行"""
    supabase = main.get_supabase()
    body = await request.json()
    code = str(body.get("code", "")).strip()

    if not code or len(code) != 6:
        raise HTTPException(status_code=400, detail="6-digit code required")

    # シークレット取得
    result = (
        supabase.table("admin_mfa")
        .select("secret_enc, enabled")
        .eq("user_id", admin_id)
        .limit(1)
        .execute()
    )
    if not result.data or not result.data[0]["enabled"]:
        raise HTTPException(status_code=404, detail="MFA not enabled")

    secret = result.data[0]["secret_enc"]
    totp = pyotp.TOTP(secret)

    if not totp.verify(code, valid_window=1):
        raise HTTPException(status_code=401, detail="Invalid code")

    # セッショントークン発行
    session = _create_session_token(supabase, admin_id)

    return {
        "status": "verified",
        **session,
    }


# ============================================================
# MFA Session Check
# ============================================================

@router.get("/session")
async def mfa_session_check(
    admin_id: str = Depends(require_admin),
    x_mfa_token: str | None = Header(None),
):
    """MFA セッショントークンの有効性を検証"""
    if not x_mfa_token:
        return {"valid": False, "reason": "no_token"}

    supabase = main.get_supabase()
    token_hash = _hash_token(x_mfa_token)
    now = datetime.now(timezone.utc).isoformat()

    result = (
        supabase.table("admin_mfa_sessions")
        .select("id, expires_at")
        .eq("user_id", admin_id)
        .eq("token_hash", token_hash)
        .gte("expires_at", now)
        .limit(1)
        .execute()
    )

    if not result.data:
        return {"valid": False, "reason": "expired_or_invalid"}

    return {
        "valid": True,
        "expires_at": result.data[0]["expires_at"],
    }

File: app/worker/src/routes/admin-mfa.ts
/**
 * /api/admin/mfa — 管理者 TOTP MFA エンドポイント
 * ポート元: app/backend/routers/admin_mfa.py
 *
 * GET  /status         MFA 設定状態
 * POST /setup          TOTP シークレット生成
 * POST /setup/verify   セットアップ確認 → セッション発行
 * POST /verify         ログイン時 TOTP 検証 → セッション発行
 * GET  /session        セッション有効性チェック
 * DELETE /session      セッション無効化
 */

import type { Env } from '../env';
import { AuthError } from '../middleware/auth';
import { requireAdmin } from '../middleware/admin-auth';
import { getSupabase } from '../lib/supabase';
import { jsonResponse, errorResponse } from '../lib/response';
import { parseJsonBody } from '../lib/validation';
import { randomBase32, verifyTotp, buildOtpUri } from '../lib/totp';
import { encryptSecret, decryptSecret } from '../lib/crypto';

const SESSION_DURATION_HOURS = 1;
const ISSUER_NAME = 'OpenRegimeAdmin';
const MAX_ATTEMPTS = 5;
const LOCKOUT_SECONDS = 900; // 15分

// インメモリ: ブルートフォース対策
const attemptTracker = new Map<string, number[]>();
// インメモリ: リプレイ防止
const usedCodes = new Map<string, Map<string, number>>();

export async function handleAdminMfa(
  request: Request,
  env: Env,
  cors: Record<string, string> | null,
): Promise<Response> {
  const supabase = getSupabase(env);
  const url = new URL(request.url);
  const path = url.pathname.replace('/api/admin/mfa', '') || '/';
  const method = request.method;

  try {
    // 全エンドポイント require_admin (MFA セッション不要、admin メールチェックのみ)
    const adminId = await requireAdmin(request, env, supabase);

    if (path === '/status' && method === 'GET') return mfaStatus(supabase, adminId, cors);
    if (path === '/setup' && method === 'POST') return mfaSetup(supabase, env, adminId, cors);
    if (path === '/setup/verify' && method === 'POST') {
      return mfaSetupVerify(request, supabase, env, adminId, cors);
    }
    if (path === '/verify' && method === 'POST') {
      return mfaVerify(request, supabase, env, adminId, cors);
    }
    if (path === '/session' && method === 'GET') {
      return mfaSessionCheck(request, supabase, adminId, cors);
    }
    if (path === '/session' && method === 'DELETE') {
      return mfaLogout(request, supabase, adminId, cors);
    }

    return errorResponse('Not Found', 404, cors);
  } catch (err) {
    if (err instanceof AuthError) return errorResponse(err.message, err.status, cors);
    return errorResponse('Internal server error', 500, cors);
  }
}

// ── ブルートフォース / リプレイ防止 ──

function checkRateLimit(userId: string): void {
  const now = Date.now() / 1000;
  const cutoff = now - LOCKOUT_SECONDS;
  const attempts = (attemptTracker.get(userId) || []).filter(t => t > cutoff);
  attemptTracker.set(userId, attempts);
  if (attempts.length >= MAX_ATTEMPTS) {
    throw new AuthError(429, `Too many attempts. Try again in ${LOCKOUT_SECONDS / 60} minutes.`);
  }
}

function recordAttempt(userId: string): void {
  const attempts = attemptTracker.get(userId) || [];
  attempts.push(Date.now() / 1000);
  attemptTracker.set(userId, attempts);
}

function clearAttempts(userId: string): void {
  attemptTracker.delete(userId);
}

function checkReplay(userId: string, code: string): void {
  const now = Date.now() / 1000;
  const userCodes = usedCodes.get(userId) || new Map();
  // 期限切れ削除
  for (const [c, exp] of userCodes) {
    if (exp < now) userCodes.delete(c);
  }
  usedCodes.set(userId, userCodes);
  if (userCodes.has(code)) {
    throw new AuthError(401, 'Code already used');
  }
}

function markCodeUsed(userId: string, code: string): void {
  const userCodes = usedCodes.get(userId) || new Map();
  userCodes.set(code, Date.now() / 1000 + 90);
  usedCodes.set(userId, userCodes);
}

// ── ヘルパー ──

async function sha256Hex(input: string): Promise<string> {
  const buf = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(input));
  return Array.from(new Uint8Array(buf)).map(b => b.toString(16).padStart(2, '0')).join('');
}

async function getUserEmail(
  supabase: ReturnType<typeof getSupabase>,
  userId: string,
): Promise<string> {
  const { data } = await supabase.table('users').select('email').eq('id', userId).limit(1);
  return data?.[0]?.email || 'admin';
}

async function createSessionToken(
  supabase: ReturnType<typeof getSupabase>,
  userId: string,
): Promise<{ token: string; expires_at: string }> {
  // crypto.randomUUID() + 追加バイトで64文字16進数トークン
  const bytes = new Uint8Array(32);
  crypto.getRandomValues(bytes);
  const token = Array.from(bytes).map(b => b.toString(16).padStart(2, '0')).join('');
  const tokenHash = await sha256Hex(token);
  const expiresAt = new Date(Date.now() + SESSION_DURATION_HOURS * 3600_000).toISOString();

  await supabase.table('admin_mfa_sessions').insert({
    user_id: userId,
    token_hash: tokenHash,
    expires_at: expiresAt,
  });

  return { token, expires_at: expiresAt };
}

// ── エンドポイント ──

// GET /api/admin/mfa/status
async function mfaStatus(
  supabase: ReturnType<typeof getSupabase>,
  adminId: string,
  cors: Record<string, string> | null,
) {
  const { data } = await supabase
    .table('admin_mfa')
    .select('enabled')
    .eq('user_id', adminId)
    .limit(1);

  if (!data?.length) {
    return jsonResponse({ mfa_enabled: false, mfa_setup: false }, 200, cors);
  }
  return jsonResponse({ mfa_enabled: data[0].enabled, mfa_setup: true }, 200, cors);
}

// POST /api/admin/mfa/setup
async function mfaSetup(
  supabase: ReturnType<typeof getSupabase>,
  env: Env,
  adminId: string,
  cors: Record<string, string> | null,
) {
  // 既存チェック
  const { data: existing } = await supabase
    .table('admin_mfa')
    .select('enabled')
    .eq('user_id', adminId)
    .limit(1);

  if (existing?.length && existing[0].enabled) {
    return errorResponse('MFA already enabled', 409, cors);
  }
  if (existing?.length && !existing[0].enabled) {
    return errorResponse('MFA setup already in progress. Complete verification or contact support.', 409, cors);
  }

  if (!env.MFA_ENCRYPTION_KEY || env.MFA_ENCRYPTION_KEY.length !== 64) {
    return errorResponse('MFA encryption not configured', 503, cors);
  }

  // TOTP シークレット生成
  const secret = randomBase32(32);
  const email = await getUserEmail(supabase, adminId);
  const provisioningUri = buildOtpUri(secret, email, ISSUER_NAME);

  // 暗号化して保存
  const encryptedSecret = await encryptSecret(secret, env.MFA_ENCRYPTION_KEY);
  await supabase.table('admin_mfa').insert({
    user_id: adminId,
    secret_enc: encryptedSecret,
    enabled: false,
  });

  // QR コードは Frontend で生成 (Worker では qrcode ライブラリ不要)
  return jsonResponse({
    secret,
    provisioning_uri: provisioningUri,
  }, 200, cors);
}

// POST /api/admin/mfa/setup/verify
async function mfaSetupVerify(
  request: Request,
  supabase: ReturnType<typeof getSupabase>,
  env: Env,
  adminId: string,
  cors: Record<string, string> | null,
) {
  const body = await parseJsonBody(request);
  const code = typeof body.code === 'string' ? body.code.trim() : String(body.code || '').trim();

  if (!code || code.length !== 6) {
    return errorResponse('6-digit code required', 400, cors);
  }

  checkRateLimit(adminId);

  const { data } = await supabase
    .table('admin_mfa')
    .select('secret_enc, enabled')
    .eq('user_id', adminId)
    .limit(1);

  if (!data?.length) return errorResponse('MFA setup not found. Call /setup first.', 404, cors);
  if (data[0].enabled) return errorResponse('MFA already enabled', 409, cors);

  // 復号 + 検証
  const secretEnc = data[0].secret_enc;
  let secret: string;
  if (secretEnc.includes(':')) {
    secret = await decryptSecret(secretEnc, env.MFA_ENCRYPTION_KEY);
  } else {
    secret = secretEnc; // レガシー平文
  }

  const valid = await verifyTotp(secret, code);
  if (!valid) {
    recordAttempt(adminId);
    return errorResponse('Invalid code', 401, cors);
  }

  checkReplay(adminId, code);
  markCodeUsed(adminId, code);
  clearAttempts(adminId);

  // MFA 有効化
  await supabase.table('admin_mfa').update({
    enabled: true,
    updated_at: new Date().toISOString(),
  }).eq('user_id', adminId);

  const session = await createSessionToken(supabase, adminId);

  return jsonResponse({ status: 'mfa_enabled', ...session }, 200, cors);
}

// POST /api/admin/mfa/verify
async function mfaVerify(
  request: Request,
  supabase: ReturnType<typeof getSupabase>,
  env: Env,
  adminId: string,
  cors: Record<string, string> | null,
) {
  const body = await parseJsonBody(request);
  const code = typeof body.code === 'string' ? body.code.trim() : String(body.code || '').trim();

  if (!code || code.length !== 6) {
    return errorResponse('6-digit code required', 400, cors);
  }

  checkRateLimit(adminId);

  const { data } = await supabase
    .table('admin_mfa')
    .select('secret_enc, enabled')
    .eq('user_id', adminId)
    .limit(1);

  if (!data?.length || !data[0].enabled) {
    return errorResponse('MFA not enabled', 404, cors);
  }

  const secretEnc = data[0].secret_enc;
  let secret: string;
  if (secretEnc.includes(':')) {
    secret = await decryptSecret(secretEnc, env.MFA_ENCRYPTION_KEY);
  } else {
    secret = secretEnc;
  }

  const valid = await verifyTotp(secret, code);
  if (!valid) {
    recordAttempt(adminId);
    return errorResponse('Invalid code', 401, cors);
  }

  checkReplay(adminId, code);
  markCodeUsed(adminId, code);
  clearAttempts(adminId);

  const session = await createSessionToken(supabase, adminId);

  return jsonResponse({ status: 'verified', ...session }, 200, cors);
}

// GET /api/admin/mfa/session
async function mfaSessionCheck(
  request: Request,
  supabase: ReturnType<typeof getSupabase>,
  adminId: string,
  cors: Record<string, string> | null,
) {
  const xMfaToken = request.headers.get('X-MFA-Token');
  if (!xMfaToken) {
    return jsonResponse({ valid: false, reason: 'no_token' }, 200, cors);
  }

  const tokenHash = await sha256Hex(xMfaToken);
  const now = new Date().toISOString();

  const { data } = await supabase
    .table('admin_mfa_sessions')
    .select('id, expires_at')
    .eq('user_id', adminId)
    .eq('token_hash', tokenHash)
    .gte('expires_at', now)
    .limit(1);

  if (!data?.length) {
    return jsonResponse({ valid: false, reason: 'expired_or_invalid' }, 200, cors);
  }

  return jsonResponse({ valid: true, expires_at: data[0].expires_at }, 200, cors);
}

// DELETE /api/admin/mfa/session
async function mfaLogout(
  request: Request,
  supabase: ReturnType<typeof getSupabase>,
  adminId: string,
  cors: Record<string, string> | null,
) {
  const xMfaToken = request.headers.get('X-MFA-Token');
  if (!xMfaToken) {
    return jsonResponse({ status: 'no_token' }, 200, cors);
  }

  const tokenHash = await sha256Hex(xMfaToken);
  await supabase
    .table('admin_mfa_sessions')
    .delete()
    .eq('user_id', adminId)
    .eq('token_hash', tokenHash);

  return jsonResponse({ status: 'logged_out' }, 200, cors);
}
